### PR TITLE
CSHARP-888 CSHARP-873 Add .NET Core 3.1, contributing guidelines, code analyzers, fix CSHARP-908 CSHARP-907

### PR DESCRIPTION
**Pull request by joao-r-reis**
Sorry for including the fixes for `CSHARP-908 - precision issue in DateRangeSerializer` and `CSHARP-907 - CPU Info empty on insights` here, they manifest under .NET Core 3.1 only so I addressed them with `CSHARP-873 - .NET Core 3.1` but I should have definitely addressed `CSHARP-888 - automated code analysis and contributing guidelines` on another PR 😞 

### Bug fixes CSHARP-907 CSHARP-908

I added the bug fixes in separate commits so you can see the bug fixes only if you look at 9d10683 (this one also includes a very small unrelated change though) and 2c3f4ae 

### Automated code analysis on CI and contributing guidelines CSHARP-888

I've set all code analysis violations that we care about to a `suggestion` severity so the builds don't fail and I created CSHARP-909 to track the fixes for them. There are thousands of these violations so it's probably a bad idea to try and fix all of them in one iteration.

The code analyzers don't run by default, they only run during the build process if the `RunCodeAnalyzers` environment variable is set (enabled on CI). This is because it slows down the build and it affects the developer experience quite a bit.

### Add .NET Core 3.1 to the test matrix (and other .NET Framework targets) CSHARP-888

For CSHARP-873, I figured it would be weird to have multiple `netcoreapp` targets since we don't have multiple .NET Framework targets so I added support for several .NET Framework targets as well. The full list of targets is:

- `net452`
- `net462`
- `net472`
- `net48`
- `netcoreapp2.0` (should probably remove this soon)
- `netcoreapp2.1`
- `netcoreapp3.1`

By default the test projects build `net462` and `netcoreapp3.1` only. To build any of the other targets the `BuildAllTargets` environment variable must be set. Also changed `Jenkinsfile` and `appveyor.yml` to add these new targets to the test matrix.

### CODE REVIEW

File: src/Cassandra.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.26403.0
+# Visual Studio Version 16
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply src/Cassandra.sln:3]: I believe that line should be kept: https://docs.microsoft.com/en-us/visualstudio/extensibility/internals/solution-dot-sln-file?view=vs-2019#file-header

[the remark this answers is not in the document]
Comment (jorgebay) [reply src/Cassandra.sln:3]: oh I see, its a meta description, sorry, nvm.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+</Project>
+```
+
+As you can see, you can set the `RunCodeAnalyzers` environment variable or you can manually edit this file locally (don't commit this change).
+
+### StyleCop
+
+We follow the style guide enforced by StyleCop but we changed a couple of the default rules:
+
+- [SA1101]: in this project we don't use the `.this` prefix for local members so we disabled [SA1101] and enabled [SX1101] instead;
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply CONTRIBUTING.md]: oops, fixed 😄

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+- `Warning`: something that the driver can recover from automatically, but indicates a configuration or programming error that should be addressed. For example: the driver connected successfully, but one of the contact points in the configuration was malformed; the same prepared statement is being prepared multiple time by the application code.
+- `Info`: something that is part of the normal operation of the driver, but might be useful to know for an operator. For example: the driver has initialized successfully and is ready to process queries; an optional dependency was detected in the classpath and activated an enhanced feature.
+
+Do not log errors that are rethrown to the client (such as the error that you're going to complete a request with). This is annoying for ops because they see a lot of stack traces that require no actual action on their part, because they're already handled by application code.
+
+The last 2 levels are for developers, to help follow what the driver is doing from a "black box" perspective (think about debugging an issue remotely, and all you have are the logs).
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply CONTRIBUTING.md]: Fixed! 
This was something that was brought over from the java driver's `CONTRIBUTING.md` file so the `2 levels` part is wrong anyway, we only have one more on C# - `Verbose`

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+### Running Integration Tests
+
+#### Required tools
+
+There are a couple of tools that you need to run the integration tests: Simulacron and CCM.
+
+##### CCM
+
+To run the integration tests you need [ccm](https://github.com/pcmanus/ccm) on your machine and make the ccm commands accessible from command line path. You should be able to run `> cmd.exe /c ccm help` using command line on Windows or `$ /usr/local/bin/ccm help` on Linux / macOS.
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply CONTRIBUTING.md]: Fixed!

**Comment by jorgebay** [comment]
:shipit:

**Comment by joao-r-reis** [comment]
Something went wrong during the windows image build, .NET Framework 4.8 dev pack is not installed apparently. I'll fix this before merging the PR.

**Pull request merged**